File: jvm-core/src/main/java/com/gxk/jvm/instruction/references/NewInst.java
package com.gxk.jvm.instruction.references;

import com.gxk.jvm.instruction.Instruction;


import com.gxk.jvm.classloader.ClassLoader;
import com.gxk.jvm.interpret.Interpreter;
import com.gxk.jvm.rtda.*;
import com.gxk.jvm.rtda.heap.Heap;
import com.gxk.jvm.rtda.heap.Class;
import com.gxk.jvm.rtda.heap.Method;
import com.gxk.jvm.rtda.heap.Instance;
import com.gxk.jvm.rtda.heap.NativeMethod;

public class NewInst implements Instruction {

  public final String clazz;

  public NewInst(String clazz) {
    this.clazz = clazz;
  }

  @Override
  public int offset() {
    return 3;
  }

  @Override
  public void execute(Frame frame) {
    Class cls = Heap.findClass(clazz);

    if (cls == null) {
      ClassLoader loader = frame.method.clazz.classLoader;
      cls = loader.loadClass(clazz);
    }

    if (cls == null) {
      throw new IllegalStateException(ClassNotFoundException.class.getName());
    }

    if (!cls.getStat()) {
      // interfaceInit
      Method cinit = cls.getClinitMethod();
      if (cinit == null) {
        cls.setStat(2);
        frame.nextPc = frame.getPc();
        return;
      }

      String clNm = cinit.nativeMethodKey();
      NativeMethod clm = Heap.findMethod(clNm);
      if (clm != null) {
        clm.invoke(frame);
      } else {
        cls.setStat(1);
        Interpreter.execute(cinit);
        cls.setStat(2);
      }
    }

    Instance obj = cls.newInstance();
    frame.pushRef(obj);
  }

  @Override
  public String format() {
    return "new " + clazz;
  }
}

File: jvm-core/src/main/java/com/gxk/jvm/instruction/references/NewInst.kt
package com.gxk.jvm.instruction.references

import com.gxk.jvm.rtda.heap.Heap.findClass
import com.gxk.jvm.instruction.Instruction
import com.gxk.jvm.rtda.heap.Heap.findNativeMethod
import com.gxk.jvm.interpret.Interpreter.execute
import com.gxk.jvm.rtda.Frame
import java.lang.ClassNotFoundException

class NewInst(val clazz: String) : Instruction {
    override fun offset(): Int {
        return 3
    }

    override fun execute(frame: Frame) {
        var cls = findClass(clazz)
        if (cls == null) {
            val loader = frame.method.clazz.classLoader
            cls = loader.loadClass(clazz)
        }
        checkNotNull(cls) { ClassNotFoundException::class.java.name }
        if (!cls.judgeStat()) {
            // interfaceInit
            val cinit = cls.clinitMethod
            if (cinit == null) {
                cls.stat = 2
                frame.nextPc = frame.pc
                return
            }
            val clNm = cinit.nativeMethodKey()
            val clm = findNativeMethod(clNm)
            if (clm != null) {
                clm.invoke(frame)
            } else {
                cls.stat = 1
                execute(cinit)
                cls.stat = 2
            }
        }
        val obj = cls.newInstance()
        frame.pushRef(obj)
    }

    override fun format(): String {
        return "new $clazz"
    }
}